// eslint-plugin-sonarjs/lib/rules/no-extra-arguments.js
"use strict";
/*
 * eslint-plugin-sonarjs
 * Copyright (C) 2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// https://jira.sonarsource.com/browse/RSPEC-930
var nodes_1 = require("../utils/nodes");
var rule = {
    create: function (context) {
        var callExpressionsToCheck = [];
        var usingArguments = new Set();
        var emptyFunctions = new Set();
        return {
            CallExpression: function (node) {
                var callExpr = node;
                if (nodes_1.isIdentifier(callExpr.callee)) {
                    var reference = context.getScope().references.find(function (ref) { return ref.identifier === callExpr.callee; });
                    var definition = reference && getSingleDefinition(reference);
                    if (definition) {
                        if (definition.type === "FunctionName") {
                            checkFunction(callExpr, definition.node);
                        }
                        else if (definition.type === "Variable") {
                            var init = definition.node.init;
                            if (init && (nodes_1.isFunctionExpression(init) || nodes_1.isArrowFunctionExpression(init))) {
                                checkFunction(callExpr, init);
                            }
                        }
                    }
                }
                else if (nodes_1.isArrowFunctionExpression(callExpr.callee) || nodes_1.isFunctionExpression(callExpr.callee)) {
                    // IIFE
                    checkFunction(callExpr, callExpr.callee);
                }
            },
            ":function": function (node) {
                var fn = node;
                if (nodes_1.isBlockStatement(fn.body) && fn.body.body.length === 0 && fn.params.length === 0) {
                    emptyFunctions.add(node);
                }
            },
            "FunctionDeclaration > BlockStatement Identifier": function (node) {
                checkArguments(node);
            },
            "FunctionExpression > BlockStatement Identifier": function (node) {
                checkArguments(node);
            },
            "Program:exit": function () {
                callExpressionsToCheck.forEach(function (_a) {
                    var callExpr = _a.callExpr, functionNode = _a.functionNode;
                    if (!usingArguments.has(functionNode) && !emptyFunctions.has(functionNode)) {
                        report(callExpr, functionNode.params.length, callExpr.arguments.length);
                    }
                });
            },
        };
        function getSingleDefinition(reference) {
            if (reference && reference.resolved) {
                var variable = reference.resolved;
                if (variable.defs.length === 1) {
                    return variable.defs[0];
                }
            }
            return undefined;
        }
        function checkArguments(identifier) {
            if (identifier.name === "arguments") {
                var reference = context.getScope().references.find(function (ref) { return ref.identifier === identifier; });
                var definition = reference && getSingleDefinition(reference);
                // special `arguments` variable has no definition
                if (!definition) {
                    var ancestors = context.getAncestors().reverse();
                    var fn = ancestors.find(function (node) { return nodes_1.isFunctionDeclaration(node) || nodes_1.isFunctionExpression(node); });
                    if (fn) {
                        usingArguments.add(fn);
                    }
                }
            }
        }
        function checkFunction(callExpr, functionNode) {
            var hasRest = functionNode.params.some(function (param) { return param.type === "RestElement"; });
            if (!hasRest) {
                if (callExpr.arguments.length > functionNode.params.length) {
                    callExpressionsToCheck.push({ callExpr: callExpr, functionNode: functionNode });
                }
            }
        }
        function report(callExpr, expected, provided) {
            // prettier-ignore
            var expectedArguments = expected === 0 ? "no arguments" :
                expected === 1 ? "1 argument" :
                    expected + " arguments";
            // prettier-ignore
            var providedArguments = provided === 0 ? "none was" :
                provided === 1 ? "1 was" :
                    provided + " were";
            var message = "This function expects " + expectedArguments + ", but " + providedArguments + " provided.";
            context.report({ message: message, node: callExpr });
        }
    },
};
module.exports = rule;
